Project Build Statuses via JS

kylekirkby opened the pull request:
This PR replaces the test results status badges on the homepage with a dynamic JS component.

danrue commented:
@kylekirkby what a nice surprise! Thank you this is really good!

@situmam do you see this? Can you review?

![image](https://user-images.githubusercontent.com/409011/71279926-9476e680-231f-11ea-8acb-d94539ce802f.png)

pcolmer commented:
@kylekirkby can anything be done to align the columns? It looks a bit untidy with the blobs not being vertically aligned.

kylekirkby commented:
> @kylekirkby can anything be done to align the columns? It looks a bit untidy with the blobs not being vertically aligned.

Hi @pcolmer,

I'm going to have a quick look now. The unordered list is using bootstrap 4 flexbox utility classes so there should be a class combination that will space items appropriately.
https://getbootstrap.com/docs/4.4/utilities/flex/

danrue commented:
Hi @kylekirkby - this looks great to me, and I would just merge it, but it takes about 30 seconds to load on my system. What can we do to make that faster?

danrue commented on assets/js/app/test_results.js:
```diff
@@ -0,0 +1,244 @@
+  {
+    name: "4.4",
+    squad_url: "https://qa-reports.linaro.org/lkft/linux-stable-rc-4.4-oe/",
+    project_url: "https://qa-reports.linaro.org/api/projects/40/"
```
Oh, please don't hardcode these here. We have _data/builds.yml. If it's missing a field, add it there please. They get updated regularly.

kylekirkby commented:
> Hi @kylekirkby - this looks great to me, and I would just merge it, but it takes about 30 seconds to load on my system. What can we do to make that faster?

Could you please see if it is any faster now? :D 

I'll remove the hard coded project url's tomorrow!

danrue commented:
That's much faster! I see you removed the "please wait" line too - thank you!

![image](https://user-images.githubusercontent.com/409011/71917079-e2c01e00-3144-11ea-8b18-224c52a05cc8.png)

mrchapp commented:
This looks great! Having that information front and center on the webpage is a great improvement.

I feel it's a little too expansive, though, like in the cell padding that makes for a really spacious table. If it could be vertically compacted (so that rows are closer together) it would feel a bit better, I believe.

danrue commented:
My punch list:
- I notice the "2 hours ago" is a link that does nothing, along with all the pass/fail/skipped bubbles.
- Instead of saying "building...", can it say "testing..."?
- If I click "View Builds" and then I click on an individual "View" link, it doesn't take me where I expect.

kylekirkby commented:
Thanks, @danrue @mrchapp for your feedback! I'll see what I can do! :)

kylekirkby commented:
> My punch list:
> 
> * I notice the "2 hours ago" is a link that does nothing, along with all the pass/fail/skipped bubbles.
> * Instead of saying "building...", can it say "testing..."?
> * If I click "View Builds" and then I click on an individual "View" link, it doesn't take me where I expect.

Thanks @danrue! I was going to take the "View link" to the CI.l.o url but collecting build metadata seems to take too long even when using deferred ajax requests so I've left it out for now. I've removed the redundant anchor's and updated the text from "building..." to "testing...". @mrchapp  I've removed the additional padding and update the progress bar so it accurately represents what's going on.

danrue commented:
Hi @kylekirkby - sorry I didn't notice this last time. Can you use the build _version_ rather than the project name in this view? e.g. `v5.5-rc5-41-gb07f636fca1c` rather than `linux-mainline-oe` in the rows below.

![image](https://user-images.githubusercontent.com/409011/72016015-86362f00-3228-11ea-8e66-fdbe81e801ea.png)

One other suggestion, and I'm just getting picky now, is: I notice that the "testing" animation adds a row to the table when it's on (makes the row twice as tall), and, is not shown on the "view builds" modal. Maybe we could not say "testing", and just add the spinner animation to the beginning of the line in both the main table, and the "view builds" table?

Anyway, thank you so much. This is really nice.

pcolmer commented:
@danrue 

> Maybe we could not say "testing"

It was my suggestion to @kylekirkby that this was added because I felt that a spinner on its own didn't explain why it was there. Is the web page still loading data, for example?

I think that there needs to be some explanation but I'm sure @kylekirkby can figure out how to get it onto the same line as the rest of the text. We do need to by mindful of smaller screens with the layout, though, so there will be some situations where things get shuffled onto subsequent lines.

kylekirkby commented:
Hi @danrue @pcolmer @mrchapp,

I believe all has been resolved now. Please review when possible!

Many thanks,

Kyle

pcolmer commented:
![image](https://user-images.githubusercontent.com/1979931/72058923-e381bd00-32c8-11ea-9372-d6e74a26a7b2.png)

Would this look better if the testing spinner were in its own column/div/span/whatever you want to call it and the test name left-justified instead of centred?

@danrue If I click on view all builds for linux-mainline-oe, all the tests are apparently still building:

![image](https://user-images.githubusercontent.com/1979931/72059101-41160980-32c9-11ea-9d56-e0e596a9d248.png)

Is that really the case or is the system misreporting?

pcolmer commented:
Nit-pick: is it possible to sort the names so that the output is always in the same order? :)

kylekirkby commented:
> ![image](https://user-images.githubusercontent.com/1979931/72058923-e381bd00-32c8-11ea-9372-d6e74a26a7b2.png)
> 
> Would this look better if the testing spinner were in its own column/div/span/whatever you want to call it and the test name left-justified instead of centred?
> 
> @danrue If I click on view all builds for linux-mainline-oe, all the tests are apparently still building:
> 
> ![image](https://user-images.githubusercontent.com/1979931/72059101-41160980-32c9-11ea-9d56-e0e596a9d248.png)
> 
> Is that really the case or is the system misreporting?

I've made these changes. 



> Nit-pick: is it possible to sort the names so that the output is always in the same order? :)

I'm looking at this now!

kylekirkby commented:
@pcolmer - _sorted_ ;)

danrue commented:
> @danrue If I click on view all builds for linux-mainline-oe, all the tests are apparently still building:
> 
> ![image](https://user-images.githubusercontent.com/1979931/72059101-41160980-32c9-11ea-9d56-e0e596a9d248.png)
> 
> Is that really the case or is the system misreporting?

@pcolmer - it may be true or it may be a bug, but either way it should be reported as shown in qa-reports. The language is a little confusing becasue it's not "building", it's "testing", but qa-reports calls everything a "build".

@kylekirkby, Sorry, I've noticed more things in "View Builds":

![image](https://user-images.githubusercontent.com/409011/72079158-0491e000-32c0-11ea-9b0c-eedccfd9a28c.png)

- order of results does not match qa-reports. It does not make sense to alpha-sort based on version. Just use build-id, descending, as qa-reports does when you click "View all Builds".
- reduce padding in numbers, increase space for 'version'
- bolder for title 'linux-stable-rc-4.14-oe'

pcolmer commented:
@danrue 

> It may be true or it may be a bug, but either way it should be reported as shown in qa-reports.

Absolutely. qa-reports **also** shows these tests/builds as running since forever :). Probably the wrong place to flag it up :)

kylekirkby commented:
> * order of results does not match qa-reports. It does not make sense to alpha-sort based on version. Just use build-id, descending, as qa-reports does when you click "View all Builds".

By build ID do you mean the `id`:

![Screenshot from 2020-01-09 16-27-34](https://user-images.githubusercontent.com/4564433/72085307-fb733400-32fc-11ea-9427-c81f0615eebb.png)

> * reduce padding in numbers, increase space for 'version'

Done :)

> * bolder for title 'linux-stable-rc-4.14-oe'

Done :)

danrue commented:
> > * order of results does not match qa-reports. It does not make sense to alpha-sort based on version. Just use build-id, descending, as qa-reports does when you click "View all Builds".
> 
> By `build-id` do you mean the `id` as listed in my results object:

Yea `id`, with the highest/newest at the top.

kylekirkby commented:
> > > * order of results does not match qa-reports. It does not make sense to alpha-sort based on version. Just use build-id, descending, as qa-reports does when you click "View all Builds".
> > 
> > 
> > By `build-id` do you mean the `id` as listed in my results object:
> 
> Yea `id`, with the highest/newest at the top.

Thanks, @danrue - committed the change.

danrue commented:
It's just so beautiful :sob:  Thanks @kylekirkby, let's merge it finally!

danrue commented:
@kylekirkby i don't really use the develop branch. Can you update the PR to merge to master?
